### PR TITLE
Modals refact - preparation for getting rid of `waitForRef`

### LINKED ISSUE #4: Improve ask for postcode/phone number
Currently we ask for these when people are replying:

image

The text below the phone number box is quite subtle, so sometimes people don't read it and therefore think their phone number would be passed on. They might therefore either be reluctant to provide it or get confused when asked for a phone number later.

We want to:

1. Remove that box from there
2. Let them reply
3. After replying, do a popup which asks for their postcode and phone number, explaining why.

### LINKED ISSUE #13: Use `useModal` composable everywhere instead of the `modal` mixin.
Now we have both a composable and a mixin with the same function.  Composable is better, so we need to migrate to that in the places where we currently use the mixin.  It doesn't need to in this PR - please create another issue for that.

_Originally posted by @edwh in https://github.com/Freegle/iznik-nuxt3/pull/8#discussion_r1304332926_

### PR DESCRIPTION
**Pull request by smellyshovel**
The main goal of this PR is to **prepare** to completely get rid of the `waitForRef`. The `waitForRef` composable is mainly used for the modals, and the main goal of this PR is to refactor all the modals to stop using the `waitForRef`. The second main goal is to get rid of the `Modal` mixin.

1. Added the `useModal.js` composable. It should be used by each modal.
2. All the modals should be imported using `defineAsyncComponent`. The visibility of a modal is controlled with the `v-if` flag. The `v-if` flag should be manually reset on the `hidden` event - that way we can respect the animations, and because of the use of `v-if` we can rely on `defineAsyncComponent` to load the modal only when it's really necessary (lazy-loading/tree-shaking).
  Example usage: `<MyModal v-if="showMyModal" @hidden="showMyModal = false" />`
3. `DonationAskModal.vue` also satisfies the points above, but also provides (and depends on) a new composable called `useDonationAskModal`. The new composable provides the flag to use in the parent and the flag is automatically set to true when appropriate. So, the flow is the following now: import the `DonationAskModal.vue` using the `defineAsyncComponent`, import the `useDonationAskModal`. Use the `DonationAskModal` as follows:
  ```
  <DonationAskModal
      v-if="showDonationAskModal"
      @hidden="showDonationAskModal = false"
    />
    
    ...
    
    const { showDonationAskModal } = useDonationAskModal()
  ```
  Thus the modal is automatically shown on a page when appropriate.
  
Checklist:
1. [x] `AboutMeModal.vue`. Used in:
    - [x] `settings/index.vue`
    - [x] `browse/[[term]].vue` - **Not sure how to reach the modal here**
    - [x] `NewsAboutMe.vue` - `NewsThread.vue` - `chitchat/[[id]].vue`
    - [x] `MainHeader.vue` - **Not sure how to reach the modal here**
2. [x] `ContactDetailsAskModal.vue`. Used in:
    - [x] `chats/[[id]].vue`
3. [x] `DonationAskModal.vue`. Used in:
    - [x] `myposts.vue` - **Checked by adding the `ask()` method call to the useDonationAskModal.js:44**
    - [x] `mypost/[id]/[[action]].vue` - **Not sure from where to reach the page**
4. [x] `ProfileModal.vue`. Used in:
    - [x] `NewsUserIntro.vue`
    - [x] `settings/index.vue`
    - [x] `ChatFooter.vue` - `ChatPane.vue` - **?**
    - [x] `ChatHeader.vue` - `ChatPane.vue` - **?**
    - [x] `MessageHistoryExpanded.vue` - `MessageExpanded.vue`
    - [x] `MyMessageReply.vue`
    - [x] `NewsReply.vue`
5. [x] `CommunityEventModal.vue`. Used in: **won't test: no events ATM, not sure whether it's OK to create**
    - [x] `communityevents/[[groupid]].vue`
    - [x] `CommunityEvent.vue`
    - [x] `CommunityFeature.vue`
    - [x] `NewsCommunityEvent.vue`
6. [x] `VolunteerOpportunityModal.vue`. Used in: **won't test: no vol. opport. ATM, not sure whether it's OK to create**
    - [x] `volunteerings/[[groupid]].vue`
    - [x] `CommunityFeature.vue`
    - [x] `NewsVolunteerOpportunity.vue`
    - [x] `VolunteerOpportunity.vue`.
7. [x] `MessageModal.vue`. Used in:
    - [x] `OurMessage.vue`
8. [x] `MessageEditModal.vue`. Used in:
    - [x] `MyMessage.vue`
9. [x] `AddressModal.vue`. Used in:
    - [x] `settings/index.vue`
    - [x] `ChatFooter.vue` - `ChatPane.vue` - **?**
10. [x] `OutcomeModal.vue`. Used in:
    - [x] `ChatMessagePromised.vue`
    - [x] `MessageEditModal.vue` - **seems like the OutcomeModal should open if the item's quantity decreases? Doesn't work even on the prod**
    - [x] `ChatMessageInterested.vue`
    - [x] `MyMessage.vue`
11. [ ] `ForgetFailModal.vue`. Used in:
    - [ ] `unsubscribe/[[id]].vue`
12. [x] `ConfirmModal.vue`. Used in:
    - [x] `ChatMessage.vue` (for delete message)
    - [x] `JoinWithConfirm.vue`
    - [ ] `NewsRefer.vue`
    - [x] `NewsReply.vue`
    - [x] `NewsThread.vue`
    - [ ] `PostPhoto.vue`
    - [ ] `SpinButton.vue`
    - [x] `unsubscribe/[[id]].vue`
13. [x] -`LoginModal.vue`. Used in:-
    - [x] `layouts/default.vue`
    - [x] `layouts/login.vue`
14. [x] `NewsShareModal.vue`. Used in: * *(still relies on the ref (see `NewsBase.vue`))*
    - [x] `NewsAlert.vue` *
    - [x] `NewsMessage.vue` *
    - [x] `NewsNoticeboard.vue` *
15. [x] `PosterModal.vue`. Used in:
    - [x] `pages/promote.vue`
    - [x] `noticeboards/[[id]].vue`
16. [x] `PromiseModal.vue`. Used in:
    - [x] `MyMessage.vue`
    - [x] `MyMessagePromisedTo.vue`
    - [x] `ChatFooter.vue` - `ChatPane.vue` - **?**
    - [x] `ChatMessageInterested.vue`
    - [x] `ChatMessagePromised.vue`
    - [x] `MyMessageReply.vue`
17. [x] `RenegeModal.vue`. Used in:
    - [x] `MyMessagePromisedTo.vue`
    - [x] `ChatMessagePromised.vue`
    - [x] `MyMessageReply.vue`
18. [x] `ResultModal.vue`. Used in:
    - [x] `ChatMessage.vue`
19. [x] `SupporterInfoModal.vue`. Used in:
    - [x] `ProfileInfo.vue`
    - [ ] `SupporterInfo.vue`
    - [x] `pages/help.vue`
20. [x] `StoryAddModal.vue`. Used in:
    - [x] `pages/stories/[[groupid]].vue`
    - [x] `pages/stories/authority/[[id]].vue`
    - [x] `pages/story/[id].vue`
    - [x] `NewsStory.vue` - **I don't know how to add a story from the /chitchat page**
21. [x] `StoryShareModal.vue`. Used in:
    - [x] `StoryOne.vue`
    - [x] `NewsStory.vue`
22. [x] `MessageReportModal.vue`. Used in:
    - [x] `MessageActions.vue`
23. [x] `MessageShareModal.vue`. Used in:
    - [ ] `MessageActions.vue`
    - [x] `MyMessage.vue`
24. [x] `MessagePhotosModal.vue`. Used in:
    - [x] `MyMessage.vue`
25. [x] `NudgeTooSoonWarningModal.vue`. Used in:
    - [x] `ChatFooter.vue` - `ChatPane.vue` - **?**
26. [x] `NudgeWarningModal.vue`. Used in:
    - [x] `ChatFooter.vue` - `ChatPane.vue` - **?**
27. [x] `ChatReportModal.vue`. Used in:
    - [x] `ChatHeader.vue` - `ChatPane.vue` - **?**
28. [x] `ChatBlockModal.vue`. Used in:
    - [x] `ChatHeader.vue` - `ChatPane.vue` - **?**
29. [ ] `ChatRSVPModal.vue`. Used in:
    - [ ] `ChatFooter.vue` - `ChatPane.vue` - **?**
30. [x] `ChatHideModal.vue`. Used in:
    - [x] `ChatHeader.vue` - `ChatPane.vue` - **?**
    - [x] `pages/chats/[[id]].vue`
31. [x] `EmailConfirmModal.vue`. Used in:
    - [x] `pages/settings/index.vue`
32. [x] `NewsEditModal.vue`. Used in:
    - [x] `NewsReply.vue`
33. [x] `NewsLovesModal.vue`. Used in:
    - [x] `NewsLoveComment.vue`
    - [x] `NewsReply.vue`
34. [x] `NewsPhotoModal.vue`. Used in: *(couldn't find no triggers)*
    - [x] `NewsStory.vue`
35. [x] `NewsReportModal.vue`. Used in:
    - [x] `NewsThread.vue`

Other tasks from the comments
 - [x] Community Events and Volunteering Opportunities: test by creating new posts
 - [x] The "Put up poster" button doesn't do anything. In the old version it also behaves strange: first opens a Share modal and then almost immidiately goes to the /promote page. What's the correct behavior?
    
All the components have been rewritten to use the composable (except for the `LoginModal.vue`). Now Requires some testing.
    
When complete, should close #4 and close #13.

### CODE REVIEW

**Comment by edwh** [comment]
@smellyshovel I've released a bunch of stuff to master so it needs a remerge at some point.

I checked out the branch to start reviewing.  The dropdown of notifications is full-width now - not sure why, but please fix so that it is the same width as before.

**Comment by smellyshovel** [comment]
@edwh must be related to the updated version of Bootstrap - forgot to mention that. Without the version update the new modals approach did not work (there was that wierd issue with 0 opacity and hidden visibility).

**Comment by edwh** [comment]
> @edwh must be related to the updated version of Bootstrap - forgot to mention that. Without the version update the new modals approach did not work (there was that wierd issue with 0 opacity and hidden visibility).

Ok.  I'll watch out for other oddities, but please find a way to fix that one.

**Comment by smellyshovel** [comment]
Here's why the menu is so wide:

Screenshot_20230927_192615

The value `1229.47px` comes from the element's inline style which is specified as `max-content`:

Screenshot_20230927_192804

Looking at the code, I wasn't able to find any place where the width was explicitly set to `max-content`, so I guess the value comes from Bootstrap (after updating its version).

If I override the original `width: min(400px, 100vw)` value with `!important`, the issue is solved:

Screenshot_20230927_193016

However, as you can see, the menu gets shifted to the right of the callee-button, while it's opened to the left in the existing version. I guess that's somehow related to the Bootstrap update again.

That can be fixed by making the menu a bit wider (e.g. 450px instead of 400px) (I guess because the menu won't fit anymore and due to that gets opened to the left of the button), but I'm not sure whether we need it. If you'd like it to be opened to the left, please, let me know. Changing the width from 400 to 450 px should be a one-liner.

**Comment by smellyshovel** [comment]
Found an issue with `MessageEditModal.vue` - after submitting the form, the modal gets closed but then is re-opened again. Investigating.

**Comment by smellyshovel** [comment]
Found the code responsible for the failure: `MessageEditModal.vue:263`. If we replace the `patch` call to something else, the issue's gone:

```
        // await this.messageStore.patch(params)
        await new Promise((resolve) => setTimeout(resolve, 1000))
```

Most probably that's due to the fact that patching the messages changes some parts of it which are used as the `key` somewhere in the messages list. Therefore, the message is changed as soon as it's patched, the Edit Modal is therefore changed as well, and when the `hide` method of the `useModal` is called, it fails silently due to the missing modal element (the previous one was replaces with a new one).

**Review by edwh — CHANGES_REQUESTED** [review]
The code changes all look pretty straightforward.   As discussed we need to do some testing.

Is it complete, though?  For example if I search for `waitForRef` I find `await this.waitForRef('newUserModal')` which seems to be using the old style.  The same for `await this.waitForRef('editModal')`.  Am I misunderstanding what is supposed to be included in this PR?

**Comment by edwh** [comment]
> Most probably that's due to the fact that patching the messages changes some parts of it which are used as the `key` somewhere in the messages list.

Yes, that's possible.  Patching will refetch the message.

You could try changing the parts of `message.js` store where it sets the message in `list`, so that for pre-existing messages, instead of overwriting the whole object (which might trigger reactivity where the `id` is used as a key) we set every property _except_ `id`.  That might stop reactivity happening.

However this might apply to anything where we use a reactive value as part of the key.  That happens a lot with lots of stores, often using `id`.  If that is the issue then we need to think about it some more.  I think normally that would just result in extra re-renders which we wouldn't notice, but your change exposes it as an issue.  

Let's see if that's the problem first.

**Comment by smellyshovel** [comment]
> The code changes all look pretty straightforward. As discussed we need to do some testing.
> 
> Is it complete, though? For example if I search for `waitForRef` I find `await this.waitForRef('newUserModal')` which seems to be using the old style. The same for `await this.waitForRef('editModal')`. Am I misunderstanding what is supposed to be included in this PR?

This PR is first of all a *preparation* for getting rid of `waitForRef`s. The respective mixin and and composable-couterpart are removed in #34. 

Moreover, I think I might have even missed some of the usages in #34 because by habit I was looking for the usages using the built-in IDE's helpers, I haven't been looking for the usages with the help of the full-text-search I think. In any case, this PR goes first. Then, before mergin the second one (#34), we'll re-review all the usages of `waitForRef`s again.

File: components/AboutMeModal.vue [hunk 1]
```diff
@@ -1,18 +1,17 @@
 <template>
   <b-modal
-    id="aboutmemodal"
-    v-model="showModal"
+    ref="modal"
     scrollable
     :title="
-      !review
+      !props.review
         ? 'Why not complete your public profile?'
```
[the remark this answers is not in the document]
Comment (smellyshovel) [reply components/AboutMeModal.vue:7]: Script setup syntax allows to define a variable `a` even if there already was a prop `a`, which is not the case with options API's `props` and `data` variables. Both usages in templates `a` and `props.a` are identical, though it's sort of a (my personal?) convention to explicitly prefix props with `props.` and don't prefix regular variables. 

Maybe there were also some eslint complaints regarding these, but I'm not sure. Generally, I believe it's a good practice, though I'm not forcing us to use it. Just my own preference.

File: pages/chats/[[id]].vue [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+      v-if="showContactDetailsAskModal"
+      @hidden="showContactDetailsAskModal = false"
+    />
+
     <div>
       <h1 class="visually-hidden">Chats</h1>
       <b-row class="m-0">
         <b-col id="chatlist" cols="12" md="4" xl="3" class="p-0 bg-white">
+          <b-button @click="ask"
+            >Show contact details {{ showContactDetailsAskModal }}</b-button
```
[the remark this answers is not in the document]
Comment (smellyshovel) [reply pages/chats/[[id]].vue]: Thanks for the reminder. Removed.

**Comment by smellyshovel** [comment]
> > Most probably that's due to the fact that patching the messages changes some parts of it which are used as the `key` somewhere in the messages list.
> 
> Yes, that's possible. Patching will refetch the message.
> 
> You could try changing the parts of `message.js` store where it sets the message in `list`, so that for pre-existing messages, instead of overwriting the whole object (which might trigger reactivity where the `id` is used as a key) we set every property _except_ `id`. That might stop reactivity happening.
> 
> However this might apply to anything where we use a reactive value as part of the key. That happens a lot with lots of stores, often using `id`. If that is the issue then we need to think about it some more. I think normally that would just result in extra re-renders which we wouldn't notice, but your change exposes it as an issue.
> 
> Let's see if that's the problem first.

OK. What I'll do then is just close the MessageEditModal *before* the `await` statement (that's while the modal for sure still exists). That works, though, obviously, not ideal. It's just that I really wouldn't like to mess around with the stores under the scope of this PR. Let's move it to another issue instead. From this PR's perspective the modal should close, and with my fix it does. The fact that it doesn't close exactly when it should is a bug and we'll address it separately.

**Comment by smellyshovel** [comment]
StoryShareModal won't work on the /stories page. Investigating...

**Comment by smellyshovel** [comment]
`NewsPhotoModal.vue` doesn't seem to work properly. Not sure if it's exactly that component, but in general when clicking photos here and there a modal with a close-up won't open.

I'm afraid that there might be a need to rewrite some of the news-related components. The obsolete `extend` syntax is used there and the code to trigger the modals relies on template refs, which should not be the case anymore, since one of the goals of this PR is to make the modals' visibility be controlled via a boolean flag.

@edwh could you please check the `NewsPhotoModal`, `NewsBase.vue` and e.g. `NewsStory.vue`? What should be the pattern for working with the modals there? Should we maybe rewrite the components to get rid of that `NewsBase` thing?

I think a similar issue also goes to the `NewsShareModal.vue`.

**Comment by edwh** [comment]
@smellyshovel Can you check whether the upgrade of bootstrap has broken the bootstrap patch please?  I think I 

> `NewsPhotoModal.vue` doesn't seem to work properly. Not sure if it's exactly that component, but in general when clicking photos here and there a modal with a close-up won't open.

Everything I try on ChitChat opens fine for me on the latest code.  For example on http://localhost:3002/chitchat/378217.

We should probably do a call/screen share on this one, but I'm busy until this evening my time.  See if you can provide a step-by-step set of instructions to repro that I can try.

**Comment by edwh** [comment]
Because of the change to where the notification drop-down opens, it gets covered by the ads.  Can you fiddle with z-indexes?

image

**Comment by edwh** [comment]
@smellyshovel I've done a bunch of testing and updated the checklist.  It's looking fine.  Let's have a call about the news aspect.

**Comment by smellyshovel** [comment]
> @smellyshovel I've done a bunch of testing and updated the checklist. It's looking fine. Let's have a call about the news aspect.

Brief agenda:

- [ ] News
- [ ] Login
- [ ] Other components I don't know how to test

**Comment by smellyshovel** [comment]
> These are ids for ChitChat threads of the different types:
http://localhost:3002/chitchat/101747
http://localhost:3002/chitchat/48093
http://localhost:3002/chitchat/378319
http://localhost:3002/chitchat/377233
http://localhost:3002/chitchat/377797
http://localhost:3002/chitchat/378406
http://localhost:3002/chitchat/358174

**Comment by smellyshovel** [comment]
Share, Love and Report, Photo (where applicable) modals work for all the example posts above.
@edwh could you please test the edit modals? I guess you should have enough permissions to be able to edit other people's posts.

 - [ ] The "Put up poster" button doesn't do anything. In the old version it also behaves strange: first opens a Share modal and then almost immidiately goes to the /promote page. What's the correct behavior?

All the rest looks good to me.

**Comment by edwh** [comment]
> * The "Put up poster" button doesn't do anything. In the old version it also behaves strange: first opens a Share modal and then almost immidiately goes to the /promote page. What's the correct behavior?

To go to /promote.  Probably it's not suppressing the event.

**Comment by edwh** [comment]
@smellyshovel Tested edit. Have you updated the checklist above for other news tests?

**Comment by smellyshovel** [comment]
@edwh now updated. So, several modals/usages are still left to check. I'll go through them once again, maybe there's something I know how/where to test. I'll notify you when I'm done with all I can do, so that you could check the rest.

**Comment by smellyshovel** [comment]
@edwh Marking this as ready for review.

I've double-checked everything that I could, though there are still some usages that you might want to check yourself before merging (see the list in the OP).

**Review by edwh — APPROVED** [review]

**Pull request merged**